[sancov] Introduce optional callback for stack-depth tracking

kees opened the pull request:
Normally -fsanitize-coverage=stack-depth inserts inline arithmetic to update thread_local __sancov_lowest_stack. To support stack depth tracking in the Linux kernel, which does not implement traditional thread_local storage, provide the option to call a function instead.

This matches the existing "stackleak" implementation that is supported in Linux via a GCC plugin. To make this coverage more performant, a minimum estimated stack depth can be chosen to enable the callback mode, skipping instrumentation of functions with smaller stacks.

With -fsanitize-coverage-stack-depth-callback-min set greater than 0, the __sanitize_cov_stack_depth() callback will be injected when the estimated stack depth is greater than or equal to the given minimum.

kees committed:
[sancov] Introduce optional callback for stack-depth tracking

Normally -fsanitize-coverage=stack-depth inserts inline arithmetic to
update thread_local __sancov_lowest_stack. To support stack depth
tracking in the Linux kernel, which does not implement traditional
thread_local storage, provide the option to call a function instead.

This matches the existing "stackleak" implementation that is supported
in Linux via a GCC plugin. To make this coverage more performant, a
minimum estimated stack depth can be chosen to enable the callback mode,
skipping instrumentation of functions with smaller stacks.

With -fsanitize-coverage-stack-depth-callback-min set greater than 0,
the __sanitize_cov_stack_depth() callback will be injected when the
estimated stack depth is greater than or equal to the given minimum.

kees commented:
> ⚠️ C/C++ code formatter, clang-format found issues in your code. ⚠️
> 
> You can test this locally with the following command:
> View the diff from clang-format here.

Whoops, yes, I've fixed these now.

melver commented on clang/include/clang/Driver/SanitizerArgs.h:
```diff
@@ -34,6 +34,7 @@ class SanitizerArgs {
   std::vector<std::string> CoverageIgnorelistFiles;
   std::vector<std::string> BinaryMetadataIgnorelistFiles;
   int CoverageFeatures = 0;
+  int StackDepthCallbackMin = 0;
```
`CoverageStackDepthCallbackMin`

kees replied:
Fixed.

melver commented on llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp:
```diff
@@ -1078,22 +1092,44 @@ void ModuleSanitizerCoverage::InjectCoverageAtBlock(Function &F, BasicBlock &BB,
+            EstimatedStackSize += TypeSize;
+          } else {
+            // Over-estimate dynamic sizes.
+            EstimatedStackSize += 4096;
```
This just seems too arbitrary. Can we do something more reliable? Because if this is used beyond the kernel (where we shouldn't even have all that many VLAs or dynamic allocas left - right?), it's anyone's guess if 4K is even a generous over estimation.

I see several options:
- Make this a cl::opt and not a hard-coded constant. The default should be much larger (8 MiB?), so that even we can't accurately calculate stack size we always get a callback.
- Just give up if we found a dynamic alloca, and always do the callback.

kees replied:
Yeah, that's a good point. And yes, we have no dynamic stack sizes in Linux anymore, so this "else" will never get run there. I think "always make the callback" is a good choice here.

melver reviewed (changes requested):
This is also missing flag and IR tests.

kees commented:
> This is also missing flag and IR tests.

Oh, yes, I will add those. Thanks!

melver reviewed (commented):
Generally LGTM - but let's also wait for others to comment.

Documentation of this feature is lacking (and afaik also wasn't added in https://reviews.llvm.org/D36839). Given this will be used in the kernel, some kind of official documentation might be good to have.

Btw, if you want a runtime test, it's possible to add one similar to the one in `compiler-rt/trunk/test/sanitizer_common/TestCases/sanitizer_coverage_stack_depth.cc`.

melver commented on clang/include/clang/Driver/Options.td:
```diff
@@ -2361,6 +2361,13 @@ def fsanitize_coverage_ignorelist : Joined<["-"], "fsanitize-coverage-ignorelist
     HelpText<"Disable sanitizer coverage instrumentation for modules and functions "
              "that match the provided special case list, even the allowed ones">,
     MarshallingInfoStringVector<CodeGenOpts<"SanitizeCoverageIgnorelistFiles">>;
+def fsanitize_coverage_stack_depth_callback_min_EQ
```
Yeah LLVM has a bunch of these hidden features, and when you find them to solve your problem it's like hitting the jackpot. ;-)

kees replied:
Documentation added.

kees committed:
[sancov] Document -fsanitize-coverage=stack-depth

Add documentation to detail the behaviors of the stack-depth tracer,
including the new -fsanitize-coverage-stack-depth-callback-min=N
argument.

kees commented:
Okay, a couple small clean-ups, and I've also solved the lack of dynamic stack size tracking.

vitalybuka commented on llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp:
```diff
@@ -1078,22 +1091,61 @@ void ModuleSanitizerCoverage::InjectCoverageAtBlock(Function &F, BasicBlock &BB,
+                      dyn_cast<ConstantInt>(AI->getArraySize()))
+                Bytes *= arraySize->getZExtValue();
+            }
+            EstimatedStackSize += Bytes;
```
why do we need to do this accounting and not just use something like __builtin_frame_address ?

kees replied:
I think you're asking why I can't use Intrinsic::frameaddress in the callback case? (This is what the `__sancov_lowest_stack` mode does.) At this point the actual size of the stack space is unknown (since it hasn't gone through CodeGen). We need to know this early to minimize the number of places where instrumentation is happening. So to estimate how much space is in use, this manually adds it up to decide whether or not to insert the callback call. (In the `__sancov_lowest_stack` mode, the instrumentation is _always_ added.)

kees replied:
I've added a more detailed comment to the accounting loop (and moved the insertion calculation there since it's only needed for the callback case). And I've tweaked the documentation a bit more with a short example.

melver commented on llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp:
```diff
@@ -1078,22 +1091,65 @@ void ModuleSanitizerCoverage::InjectCoverageAtBlock(Function &F, BasicBlock &BB,
+      // If dynamic alloca found, always add call.
+      bool dynamic_alloca = false;
+      // Find an insertion point after last "alloca".
+      llvm::Instruction *InsertBefore = NULL;
```
nullptr

melver commented on llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp:
```diff
@@ -1078,22 +1091,65 @@ void ModuleSanitizerCoverage::InjectCoverageAtBlock(Function &F, BasicBlock &BB,
+      // In callback mode, only add call when stack depth reaches minimum.
+      uint32_t EstimatedStackSize = 0;
+      // If dynamic alloca found, always add call.
+      bool dynamic_alloca = false;
```
HasDynamicAlloc?

kees replied:
Fixed

melver commented on clang/docs/SanitizerCoverage.rst:
```diff
@@ -385,6 +385,49 @@ Users need to implement a single function to capture the CF table at startup:
+
+The maximum depth of a function call graph is stored in the thread-local
+``__sancov_lowest_stack`` variable. Instrumentation is inserted in every
+non-leaf function to check the stack pointer against this variable,
```
You write "stack pointer" here, but in the below it's "frame pointer", which are different.

melver commented on clang/docs/SanitizerCoverage.rst:
```diff
@@ -385,6 +385,49 @@ Users need to implement a single function to capture the CF table at startup:
+  if (stack < __sancov_lowest_stack)
+    __sancov_lowest_stack = stack;
+
+If ``-fsanitize-coverage-stack-depth-callback-min=N`` is also used, the
```
"If ``-fsanitize-coverage-stack-depth-callback-min=N`` (where ``N`` > 0) is also used ..." ?

Because right now it's unclear what behaviour is supposed to be if N==0. Per current implementation N==0 disables this feature.

kees replied:
That's a good clarification. Added.

melver commented on clang/docs/SanitizerCoverage.rst:
```diff
@@ -385,6 +385,49 @@ Users need to implement a single function to capture the CF table at startup:
+
+.. code-block:: c++
+
+  thread_local uintptr_t __sancov_lowest_stack;
```
extern ?

Because the compiler doesn't define the TLS variable.

kees replied:
Ah, yes!

melver commented on clang/docs/SanitizerCoverage.rst:
```diff
@@ -385,6 +385,49 @@ Users need to implement a single function to capture the CF table at startup:
+case of a dynamically sized stack, the callback is unconditionally added.
+
+The callback takes no arguments and is responsible for determining the
+stack pointer and doing any needed comparisons and storage. A roughtly
```
s/roughtly/roughly/

bwendling commented on llvm/include/llvm/Transforms/Utils/Instrumentation.h:
```diff
@@ -158,6 +158,7 @@ struct SanitizerCoverageOptions {
   bool PCTable = false;
   bool NoPrune = false;
   bool StackDepth = false;
+  int StackDepthCallbackMin = 0;
```
Could it make sense to put this at the end of the `bool` list? I wonder if it would make the struct size smaller or not. TBH, this struct should probably use `llvm/include/llvm/ADT/Bitfields.h` for the booleans.

kees replied:
Yeah, I wasn't sure if I should keep it with the other StackDepth var. I've moved it now.

bwendling commented on llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp:
```diff
@@ -1078,22 +1091,65 @@ void ModuleSanitizerCoverage::InjectCoverageAtBlock(Function &F, BasicBlock &BB,
+      for (auto &I : BB) {
+        if (auto *AI = dyn_cast<AllocaInst>(&I)) {
+          // Move potential insertion point past the "alloca".
+          InsertBefore = I.getNextNode();
```
nit: I normally use the `AI` in situations like this instead of the pre-casted value.

kees replied:
Ah, yeah, good point. Fixed.
